---
title: "R_Project"
author: "Josemari"
date: "11/27/2020"
output: html_document
---

```{R}
# Load libraries.
library(DESeq2)
library(ggplot2)
library(reshape2)
```

```{r}
# Functions.

# Write a function that returns a volcano plot for a given a dataset.
# It requires a dataset and the desired name for the plot as inputs.
volcano <- function(dataset, name){
  
  # Plot log2Fold in the x-axis and pvalue on y-axis to make a basic volcano plot.
  with(dataset, plot(log2FoldChange, -log10(pvalue), pch = 20, main = name, xlim = c(-4, 4)))
  
  # Subset the dataset and add colors to highlight important genes.
  # blue: padj < 0.01.
  with(subset(dataset, padj < 0.01 ), points(log2FoldChange,
                                             -log10(pvalue),
                                             pch = 20,
                                             col = "blue"))
  # red: log2FC > 2 and padj < 0.01).
  with(subset(dataset, padj < 0.01 & abs(log2FoldChange) > 2), points(log2FoldChange,
                                                                      -log10(pvalue),
                                                                      pch = 20,
                                                                      col = "red"))
  
  return()
}

# Write a function that returns the differentially expressed genes from a dataset.
# It requires the dataset and the desired pvalue and foldchange thresholds as inputs.
get_DE_genes <- function(dataset, foldchange_threshold, p_value_threshold){
  
  # Set a filter being logfold absolute values higher than the threshold and p value lower than threshold.
  filter = which(abs(dataset$log2FoldChange) > foldchange_threshold & dataset$padj < p_value_threshold)
  
  # Take from the dataset only the rows that fit the condition.
  DE_genes = dataset[filter,]
  
  # Get the indices of the rows sorted decreasingly using the logfold.
  index = sort.int(DE_genes$log2FoldChange, decreasing = TRUE, index.return = TRUE)$ix
  
  # Sort the genes according to the index to get the upregulated genes  at the top.
  DE_genes = DE_genes[index,]
  
  return(DE_genes)
}
```

```{r}
# Load the dataset maintaining the row names and column names.
data <- read.table("RProject_Data.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE, row.names = 1)
# Substract the common tag from the gene names and leave only the locus_tag, which is unique to each gene.
rownames(data) = sapply(rownames(data), function(x) unlist(strsplit(x, "_"))[2])
# Load the file with the experiment design to identify each count column with its sample type.
samples <- read.table("RProject_Data_ID.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
# Have a look at the data.
head(data)
```

```{r}
# Use the DESeq2 R package.
# Create the DESeq object.
dds <- DESeqDataSetFromMatrix(countData = data, colData = samples, design =~ Design)
# Run the DESeg2 function on the DESeq object.
dds <- DESeq(dds)
# Have a look at the object.
dds
```

```{r}
# Run the variant stabilizing transformation to normalize the data.
vsdata <- vst(dds, blind = FALSE)
# Rename the columns with the experiment design instead of the bam file names.
colnames(vsdata) <- samples$Design
# Make a PCA plot to visualize the variance across the samples.
plotPCA(vsdata, intgroup = "Design")
```

```{r}
# Use the results function to compare the gene counts of the mutant with the wild type.
dataResults <- results(dds, contrast = c("Design", "KO", "WT"))
head(dataResults)
```

```{r}
# Make a volcano plot for the results that will plot all the differentially expressed genes.
volcano(dataResults, "Volcano Plot: KO vs WT")
```

```{r}
# Get only the most differentially expressed genes using the get_DE_genes() function.
# Use fold change threshold of 3 and pvalue threshold of 0.01.
DE_genes = get_DE_genes(dataResults, 3, 0.01)

# Convert the vsdata into a dataframe and get the rownames in a variable.
vsDataFrame <- as.data.frame(assay(vsdata))
vsDataFrame$locus_tag <- rownames(vsDataFrame)

# From the results df, get only the genes that were over the threshold conditions we set up previously.
DE_genes_result = vsDataFrame[rownames(DE_genes), c(1:3, 4:6, 7)]
DE_genes_result
```

```{r}
par(mfrow = c(3, 3))
plotCounts(dds, gene = "MB1832c", intgroup = "Design")
plotCounts(dds, gene = "MB1554c", intgroup = "Design")
plotCounts(dds, gene = "MB1544", intgroup = "Design")
plotCounts(dds, gene = "MB2417B", intgroup = "Design")
plotCounts(dds, gene = "MB2417A", intgroup = "Design")
plotCounts(dds, gene = "MB0780", intgroup = "Design")
```

```{r}
# Final Heatmap.

# Reformat the data temporal DE data in an easy way to read for the heatmap.
GenesHeatmap <- melt(DE_genes_result, id.vars = c("locus_tag"))

# Make the geneID rows reorganized so the upregulated genes end up on top.
GenesHeatmap$locus_tag <- factor(GenesHeatmap$locus_tag, levels = rev(rownames(DE_genes)))

# Plot the heatmap.
ggplot(GenesHeatmap, aes(x = variable, y = locus_tag, fill = value)) + geom_tile() + xlab("Samples") + ylab("locus_tag") + ggtitle("KO vs WT") + theme(legend.title = element_text(size = 9), plot.title = element_text(size = 12), axis.title = element_text(size = 10, face = "bold"), axis.text.y = element_text(size = 8), axis.text.x = element_text(angle = 65, hjust = 1)) + labs(fill = "Reads")

```
